### PR TITLE
Chapter9

### PR DESCRIPTION
**Pull request by danieljharris**
"I'm not sure what Exercise 9-6 is asking me to do, when your free could you take a look at it for me and see if you understand what its asking"

### CODE REVIEW

**Comment by velimir** [comment]
in 9-2, as to my mind, they ask you to serialise binary tree using some homegrown binary format:
`[8,6,2,cat,2,dog,emu,fish]` to binary. 

page 210:
> Once you have a list representation, you can convert it into a stream of bytes in a straightforward way. You also can remove the intermediate lists from the construction entirely, going directly from tree to bitstream, and vice versa.

You need to think how you can represent atoms in binary format and convert format (see that list I mentioned) into bitstream, which can be then deserialised back to erlang list (which in turn can be deserialised into binary tree).

**Comment by danieljharris** [comment]
I've updated the "trees" module with a solution to 9-6, it works but the question says "You will need to think about how much size is needed", and I haven't, maybe I'm doing it a not efficient way?

File: Chapter9/hof.erl [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+% 		_ ->
+% 			filter(P,Xs)
+% 	end.
+
+
+print_fun() ->
+	fun(Val) -> io:format("Integers: ~p~n", [Val]) end.
+
+print(Start, End) ->
+	foreach(print_fun(), lists:seq(Start, End)).
```
Comment (velimir) [inline Chapter9/hof.erl]: You can pass a function as `fun print_fun/0`. In this case `F` variable (from `foreach/2`)  is bound to `print_fun` itself.

File: Chapter9/hof.erl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+% 	case P(X) of
+% 		true ->
+% 			[X| filter(P,Xs)];
+% 		_ ->
+% 			filter(P,Xs)
+% 	end.
+
+
+print_fun() ->
+	fun(Val) -> io:format("Integers: ~p~n", [Val]) end.
```
Comment (velimir) [inline Chapter9/hof.erl]: You could've used this function directly: 
```erlang
print(Val) ->
    io:format("Integers: ~p~n", [Val]).

print(Start, End) ->
    foreach(fun print/1, lists:seq(Start, End)).
```

File: Chapter9/hof.erl [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	foreach(print_fun(), Even).
+
+
+smaller_or_equal(List, Max) ->
+	lists:filter(fun(A) -> A =< Max end, List).
+
+
+% Not sure if this is how they wanted me to structure the function
+% but its the best way I could get working that involved funs
+concat() -> fun(List) -> lists:concat(List) end.
```
Comment (velimir) [inline Chapter9/hof.erl]: I assume they what you to create an append function, that for a given a list of lists it returns back a list, which consists of elements from the given lists.
Something like:
```erlang
hof:concat([[1,2,3], [foo, bar, baz], ["derp"]]). %% => [1,2,3, foo, bar, baz, "derp"]
```

Please [note](http://erlang.org/doc/man/lists.html#concat-1) that `lists:concat/1` also converts elements to lists:
```erlang
> listsconcat([1,2,3, foo, bar, baz, "derp"]).
"123foobarbazderp"
```

File: Chapter9/ziper.erl [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+
+% "Using this example, define the function zipWith that applies a
+% binary function to two lists of arguments, in lock step:"
+
+% Not sure what the question is asking for, but I made this function
+% that works as shown
+zipWith([],_) -> [];
+zipWith(_,[]) -> [];
+zipWith([H1|T1],[H2|T2]) -> [H1 + H2 | zipWith(T1, T2) ].
```
Comment (velimir) [inline Chapter9/ziper.erl]: Your implementation has a different interface. They ask for extensible function, where you can provide function what to do with zipped elements:
```erlang
-spec zipWith(
        Fun :: fun((LH :: any(), RH::any()) ->any()),
        List1 :: list(),
        List2 :: list()
       ) -> list().
```

Your implementation can only sum elements.

**Review by velimir — CHANGES_REQUESTED** [review]
regarding your question about exercise 9-6: I assume that they wanted you to store info about list size, atom sizes and so on in a binary format, though your solution also works. It's fine to leave it as is.

File: Chapter9/hof.erl (hunk 2 above)
Comment (danieljharris) [reply Chapter9/hof.erl]: Ahh ok, yeah thats better than how I did it, just updated the code

File: Chapter9/hof.erl (hunk 3 above)
Comment (danieljharris) [reply Chapter9/hof.erl]: Just updated the code so it takes a list of lists and using lists:foldl/3 and funs it appends the lists

File: Chapter9/ziper.erl (hunk 4 above)
Comment (danieljharris) [reply Chapter9/ziper.erl]: Updated zipWith so it now works using a passed function

File: Chapter9/hof.erl (hunk 1 above)
Comment (danieljharris) [reply Chapter9/hof.erl]: Updated code

**Review by velimir — APPROVED** [review]

**Pull request merged**